### ISSUE TITLE
OpenID Connect with Azure AD B2C and no access_token

### ISSUE THREAD

**Issue by frosttho** [cirrusidentity/simplesamlphp-module-authoauth2#92]
Dear all,

thank you for providing this module. It brings me a lot closer to where I want to end up. I am currently trying to connect [Azure AD B2C](https://learn.microsoft.com/en-us/azure/active-directory-b2c/overview) as an OpenID Connect identity provider to SimpleSAMLphp.

I got multiple partial solutions to work partially but never got far. The key error is probably that Azure AD B2C apparently does not provide an access_token or a userinfo endpoint by default (implicit flow) because all necessary information is already included in the id_token.

Thus, my key question is whether it is possible to configure an auth source which does not need a userinfo endpoint and also does not require an access_token to be sent with the id_token.

Would be great to hear back from you about this. I already checked out [thenetworg/oauth2-azure](https://github.com/thenetworg/oauth2-azure) and [steverhoades/oauth2-openid-connect-client](https://github.com/steverhoades/oauth2-openid-connect-client) but neither of them seemed to work for my scenario.

Thank you!

**Comment by pradtke** [comment]
I think all the authsources in the module use the `authorization code` flow and assume there is a userinfo endpoint to call.
I suppose you could modify one of them to make calling the user info endpoint something that could be disabled.

**Comment by frosttho** [comment]
Thank you for your answer, @pradtke! I did now something like that.

For everyone else looking for similar solutions - this is my workaround:

1. Satisfy the check for urlResourceOwnerDetails by adding it with an empty string in authsources.php: `'urlResourceOwnerDetails' => ''`
2. I added a new MicrosoftB2CAuth.php by copying src/Auth/Source/MicrosoftHybridAuth.php and added a new `finalStep` function by copying it from src/Auth/Source/OAuth2.php and deleting the part that uses the extracted access token.
3. However, this only stops **using** the access token. Not checking for its existence.
4. To satisfy the check for an access_token, Azure AD B2C can provide a "useless" token if a custom scope is added via the B2C configuration (see [here](https://learn.microsoft.com/en-us/azure/active-directory-b2c/add-web-api-application?tabs=app-reg-ga#configure-scopes)). This access token then should grant access to a non-existing API, fulfilling the existence-check.
5. Add the required fields from the `id_token` in the `postFinalStep `function, copying the two already existing fields.
